### PR TITLE
Add Complex Number type to GPDB

### PR DESCRIPTION
**Pull request by atris**

### CODE REVIEW

**Comment by ivannovick** [comment]
@mkiyama this will need to be documented in 5.0 that GPDB now supports complex numbers.

File: src/backend/cdb/cdbhash.c [hunk 1]
```diff
@@ -572,6 +578,29 @@ hashDatum(Datum datum, Oid type, datumHashFunction hashFn, void *clientData)
...
+			complex_real = re(complex_ptr);
+			complex_imag = im(complex_ptr);
+			/*
+			* On IEEE-float machines, minus zero and zero have different bit
+			* patterns but should compare as equal.  We must ensure that they
+			* have the same hash value, which is most easily done this way:
+			*/
+			if (0 == complex_real)
+			{
+				complex_real = (double) 0;
```
Comment (kuien) [inline src/backend/cdb/cdbhash.c]: double equals float8 here? If yes, seems the latter is easy to understand.

File: src/backend/utils/adt/complex_type.c [hunk 2]
```diff
@@ -0,0 +1,1361 @@
...
+	/* skip trailing whitespaces */
+	while ('\0' != *num && isspace(*num))
+	{
+		num++;
+	}
+
+	if ('\0' != *num)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_TEXT_REPRESENTATION),
+		   errmsg("invalid input syntax for type complex: \"%s\"", orig_num),
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: better to indent this as line 234

File: src/backend/utils/adt/complex_type.c [hunk 3]
```diff
@@ -0,0 +1,1361 @@
...
+	double		re = 0;
+	double		im = 0;
+	int			im_sign = 1;
+	char	   *orig_num = num;
+	Complex    *comp = (Complex *) palloc(sizeof(Complex));
+
+	if ('\0' == *num)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_TEXT_REPRESENTATION),
+		   errmsg("invalid input syntax for type complex: \"%s\"", orig_num),
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: better to indent this as line 234

File: src/backend/utils/adt/complex_type.c [hunk 4]
```diff
@@ -0,0 +1,1361 @@
...
+			{
+				/* store the sign of the imaginary with the imaginary part */
+				im = im * im_sign;
+				num++;
+			}
+			else
+			{
+				ereport(ERROR, (errcode(ERRCODE_INVALID_TEXT_REPRESENTATION),
+								errmsg("invalid input syntax for type complex: \"%s\"", orig_num),
+								errOmitLocation(true)));
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: This section appears 3 times with uniform format, may we consider to declare it with a Macro?

File: src/backend/utils/adt/complex_type.c [hunk 5]
```diff
@@ -0,0 +1,1361 @@
...
+			(*p) += snprintf((*p), limit - (*p), "%s", "-Infinity");
+			break;
+		default:
+			if (isnan(part))
+			{
+				(*p) += snprintf((*p), limit - (*p), "%s", "NaN");
+			}
+			else
+			{
+				int			ndig = DBL_DIG + extra_float_digits;
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: is indent between 'int' and 'ndig' too long?

File: src/backend/utils/adt/complex_type.c [hunk 6]
```diff
@@ -0,0 +1,1361 @@
...
+{
+	Complex    *num = PG_GETARG_COMPLEX_P(0);
+	char	   *ascii = (char *) palloc(MAXCOMPLEXWIDTH + 1);
+	char	   *limit = ascii + MAXCOMPLEXWIDTH + 1;
+	char	   *p = ascii;
+
+	*p = '\0';
+	complex_part_out(&p, limit, re(num));
+	p += snprintf(p, limit - p, " %c ", (im(num) >= 0 || isnan(im(num))) ? '+' : '-');
+	/*
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: It would be better if we add an empty line before a multiple-line comment.

File: src/backend/utils/adt/complex_type.c [hunk 7]
```diff
@@ -0,0 +1,1361 @@
...
+	key.y = im(num);
+
+	/*
+	 * On IEEE-float machines, minus zero and zero have different bit patterns
+	 * but should compare as equal.  We must ensure that they have the same
+	 * hash value, which is most easily done this way:
+	 */
+	if (0 == key.x)
+	{
+		key.x = 0;
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: should be '(double) 0' or '(float8) 0'?

File: src/backend/utils/adt/complex_type.c [hunk 8]
```diff
@@ -0,0 +1,1361 @@
...
+	}
+	if (0 == key.y)
+	{
+		key.y = 0;
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: should be '(double) 0' or '(float8) 0'?

File: src/backend/utils/adt/complex_type.c [hunk 9]
```diff
@@ -0,0 +1,1361 @@
...
+/*
+ *		complex_div				- returns arg1/arg2
+ */
+Datum
+complex_div(PG_FUNCTION_ARGS)
+{
+	Complex    *a = PG_GETARG_COMPLEX_P(0);
+	Complex    *b = PG_GETARG_COMPLEX_P(1);
+	Complex    *c = (Complex *) palloc(sizeof(Complex));
+	double		div = re(b) * re(b) + im(b) * im(b);
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: may we suffer from the error of 'division by zero'?

Or, for double, 'inf' is acceptable, then for complex, is 'inf' acceptable as well?

File: src/backend/utils/adt/complex_type.c [hunk 10]
```diff
@@ -0,0 +1,1361 @@
...
+					a,
+					b;
+
+		/* Let r = hypot(re(&x), im(&x)); */
+		r = DatumGetFloat8(DirectFunctionCall1(complex_mag, ComplexPGetDatum(&x)));
+		/* Let i = atan2(im(&x), re(&x)); */
+		i = DatumGetFloat8(DirectFunctionCall1(complex_arg, ComplexPGetDatum(&x)));
+		/* Let theta = i * yr; */
+		theta = dp_mul(i, yr);
+		if (yi == 0.0)
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: better to add { }

File: src/backend/utils/adt/complex_type.c [hunk 11]
```diff
@@ -0,0 +1,1361 @@
...
+complex_pow(PG_FUNCTION_ARGS)
+{
+	Complex    *a = PG_GETARG_COMPLEX_P(0);
+	Complex    *b = PG_GETARG_COMPLEX_P(1);
+	Complex    *c = (Complex *) palloc(sizeof(Complex));
+
+	/* We must check overflow each step, so we can't use cpow directly */
+	*c = pg_cpow(*a, *b);
+
+	PG_RETURN_COMPLEX_P(c);
```
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: should we call CHECKFLOATVAL() to check the overflow/underflow here? pg_cpow() doesn't do this check.

File: src/test/regress/bugbuster/known_good_schedule [hunk 12]
```diff
@@ -1,4 +1,4 @@
-##
+#
```
Comment (kuien) [inline src/test/regress/bugbuster/known_good_schedule]: what's happened here?

**Comment by kuien** [comment]
should we update the catversion.h as well as 4.3.json? as several files pg_xx.h have been updated.

File: src/test/regress/bugbuster/known_good_schedule (hunk 12 above)
Comment (atris) [inline src/test/regress/bugbuster/known_good_schedule]: A removed test. Will fix.

**Comment by atris** [comment]
Forgot to add the cat version update,thanks for bringing that to notice.

File: src/backend/utils/adt/complex_type.c (hunk 9 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: I think inf should be acceptable here.

**Comment by yydzero** [comment]
@macroyuyang Would you please take a look? Thanks!

**Comment by atris** [comment]
Fixed the cat version,  thanks!

File: src/test/regress/bugbuster/known_good_schedule (hunk 12 above)
Comment (atris) [inline src/test/regress/bugbuster/known_good_schedule]: Reverted the change since the test shouldn't be a part of bugbuster but ICG

File: src/backend/utils/adt/complex_type.c (hunk 9 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: I think it is. I saw no literature which explicitly defined what to do in
this case so I think we can go with this for now

File: src/backend/utils/adt/complex_type.c (hunk 7 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: Good catch, thanks!

File: src/backend/utils/adt/complex_type.c (hunk 4 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: No,I do not think so, since that leads to reduced readability IMO. Good catch though

File: src/backend/utils/adt/complex_type.c (hunk 8 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: Duplicate comment?

File: src/backend/utils/adt/complex_type.c (hunk 11 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: I am not sure if I follow you here. pg_cpow does call CHECKFLOATVAL for both parts of the number.

File: src/backend/utils/adt/complex_type.c (hunk 10 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: Fixed

File: src/backend/utils/adt/complex_type.c (hunk 5 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: Fixed

File: src/backend/utils/adt/complex_type.c (hunk 6 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: Fixed

File: src/backend/utils/adt/complex_type.c (hunk 7 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: Fixed

**Comment by atris** [comment]
@kuien I fixed all the above comments. Please have a look

File: gpMgmt/bin/gppylib/data/4.3.json [hunk 13]
```diff
@@ -1,7 +1,7 @@
 {
-   "__comment" : "Generated by tidycat.pl version 34 on Tue Mar 22 11:22:32 2016 CATALOG_VERSION_NO=300705112",
+   "__comment" : "Generated by tidycat.pl version 34 on Thu Apr  7 10:15:45 2016 CATALOG_VERSION_NO=32016079",
    "__info" : {
-      "CATALOG_VERSION_NO" : "300705112"
+      "CATALOG_VERSION_NO" : "32016079"
```
Comment (kuien) [inline gpMgmt/bin/gppylib/data/4.3.json]: We have a flexible policy to generate the catalog version with a simple guide pattern: 3yyymmddn. From this point, your catversion is shorter by 1 digital. A possible example is : 301604121?

File: src/backend/cdb/cdbhash.c [hunk 14]
```diff
@@ -32,6 +32,7 @@
 #include "utils/fmgroids.h"
 #include "utils/lsyscache.h"
 #include "utils/syscache.h"
+#include "utils/complex_type.h"
```
Comment (kuien) [inline src/backend/cdb/cdbhash.c]: I have no objection here. 

Yesterday I discussd with Yandong and other teammates on the sort of header files, we tend to choose alphabetic order if no other considerations. FYI.

File: src/backend/utils/adt/complex_type.c (hunk 11 above)
Comment (kuien) [inline src/backend/utils/adt/complex_type.c]: OK, if pg_cpow() does this, we surely needn't check it repeatedly.

File: src/include/catalog/catversion.h [hunk 15]
```diff
@@ -56,6 +56,6 @@
  */
 
 /*							3yyymmddN */
-#define CATALOG_VERSION_NO	300705112
+#define CATALOG_VERSION_NO	32016079
```
Comment (kuien) [inline src/include/catalog/catversion.h]: consider the version generation pattern (commented on json file), we'd better remain the same.

**Comment by atris** [comment]
@kuien I fixed your last two comments as well. Please have a look

File: src/include/catalog/pg_opfamily.h [hunk 16]
```diff
@@ -159,6 +159,8 @@ DATA(insert OID = 2745 (	2742	array_ops		PGNSP PGUID ));
 DATA(insert OID = 2968 (	403		uuid_ops		PGNSP PGUID ));
 DATA(insert OID = 2969 (	405		uuid_ops		PGNSP PGUID ));
 
+/* Complex Number type */
+DATA(insert OID = 2003 (	403		complex_ops		PGNSP PGUID ));
```
Comment (hlinnaka) [inline src/include/catalog/pg_opfamily.h]: Later versions of PostgreSQL use OID 2003 for 'textanycat' function. Would be good to find an OID that's not used in upstream either, otherwise we'll have to change this later as we merge with later versions.

**Comment by atris** [comment]
@hlinnaka Fixed OID and downgraded casts. Please see

File: src/test/regress/bugbuster/expected/gpsort.out [hunk 17]
```diff
@@ -1241,23 +1241,15 @@ LINE 1: select col19 from gpsort_alltypes order by col19 desc;
                                                    ^
 HINT:  Use an explicit ordering operator or modify the query.
 select col20 from gpsort_alltypes order by col20 asc;
- col20
```
Comment (kuien) [inline src/test/regress/bugbuster/expected/gpsort.out]: Sorry if I am misled, what's wrong with these tests? I notice you just alternate the OID, :)

Comment (atris) [inline src/test/regress/bugbuster/expected/gpsort.out]: Tests were originally changed because Point could now be sorted with implicit casting to Complex but per Heikki's advise, I downgraded Complex to Point and Point to Complex casts to explicit and hence the original tests became valid again hence reverted the test changes.
OID change is for avoiding conflict with upcoming PG 8.3 merge.

Comment (kuien) [inline src/test/regress/bugbuster/expected/gpsort.out]: Got it, thanks

**Comment by kuien** [comment]
LGTM.

File: src/backend/cdb/cdbhash.c [hunk 18]
```diff
@@ -572,6 +578,29 @@ hashDatum(Datum datum, Oid type, datumHashFunction hashFn, void *clientData)
...
+		case COMPLEXOID:		
+			complex_ptr  = DatumGetComplexP(datum);
+			complex_real = re(complex_ptr);
+			complex_imag = im(complex_ptr);
+			/*
+			* On IEEE-float machines, minus zero and zero have different bit
+			* patterns but should compare as equal.  We must ensure that they
+			* have the same hash value, which is most easily done this way:
+			*/
+			if (0 == complex_real)
```
Comment (hlinnaka) [inline src/backend/cdb/cdbhash.c]: I'd suggest copying this code more exactly from float8. I.e. I'd suggest wording this as:

```
 if (complex_real == (double) 0)
  complex_real = 0.0;
```

I think it compiles to the same, but making the source look more identical would make it more clear that it's supposed to do the exact same thing.

File: src/backend/utils/adt/complex_type.c [hunk 19]
```diff
@@ -0,0 +1,1385 @@
+ /*-------------------------------------------------------------------------
+ *
+ * complex.c
```
Comment (hlinnaka) [inline src/backend/utils/adt/complex_type.c]: typo: the file is called complex_type.c

Comment (hlinnaka) [inline src/backend/utils/adt/complex_type.c]: also remove the extra space at the very beginning.

File: src/backend/utils/adt/complex_type.c [hunk 20]
```diff
@@ -0,0 +1,1385 @@
...
+		{
+			val = -get_float8_infinity();
+			end_ptr += 9;
+		}
+		else if (ERANGE == errno)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+					 errmsg("\"%s\" is out of range for type double precision", orig_num),
+					 errOmitLocation(true)));
```
Comment (hlinnaka) [inline src/backend/utils/adt/complex_type.c]: The errOmitLocation(true) calls are not needed. Providing an error code other than ERRCODE_INTERNAL_ERROR does that implicitly.

File: src/backend/utils/adt/complex_type.c [hunk 21]
```diff
@@ -0,0 +1,1385 @@
...
+						errmsg("invalid input syntax for type complex: \"%s\"", orig_num),
+						errOmitLocation(true)));
+	}
+
+	INIT_COMPLEX(comp, re, im);
+	PG_RETURN_COMPLEX_P(comp);
+}
+
+/* max size of the representation of a complex number */
+#define MAXCOMPLEXWIDTH 320
```
Comment (hlinnaka) [inline src/backend/utils/adt/complex_type.c]: How is this calculated?

File: src/test/regress/greenplum_schedule [hunk 22]
```diff
@@ -130,4 +130,5 @@ test: uaocs_compaction/index_stats
 test: uaocs_compaction/index
 test: uaocs_compaction/drop_column
 
+test: complex
```
Comment (hlinnaka) [inline src/test/regress/greenplum_schedule]: I'd suggest putting this in one of the earlier parallel groups early in the schedule. The earlier tests are more similar to this one than the uaocs_* tests.

File: src/include/utils/complex_type.h [hunk 23]
```diff
@@ -0,0 +1,198 @@
+/*-------------------------------------------------------------------------
+ *
+ * complex.h - Declarations for complex data type
```
Comment (hlinnaka) [inline src/include/utils/complex_type.h]: typo: the file is called complex_type.h

File: src/include/utils/complex_type.h [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+ *
+ * Portions Copyright (c) 1996-2008, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ * Portions Copyright (c) 2016, Pivotal Software Inc
+ *
+ *
+ * NOTE
+ *	  These routines do *not* use the float types from adt/.
+ *
+ *	  XXX These routines were not written by a numerical analyst.
```
Comment (hlinnaka) [inline src/include/utils/complex_type.h]: LOL. Perhaps we should run these past a numerical analyst, though? We might have some in the company, or at least know someone.

File: src/include/utils/complex_type.h [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+ *
+ *-------------------------------------------------------------------------
+ */
+#ifndef COMPLEX_H
+#define COMPLEX_H
+#include "postgres.h"
```
Comment (hlinnaka) [inline src/include/utils/complex_type.h]: Headers should not include postgres.h.

File: src/include/utils/complex_type.h [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+#include "fmgr.h"
+#include <complex.h>
```
Comment (hlinnaka) [inline src/include/utils/complex_type.h]: Is this needed? Does this use any routines from the system complex.h header?

File: src/include/utils/complex_type.h [hunk 27]
```diff
@@ -0,0 +1,198 @@
...
+ *
+ *
+ * NOTE
+ *	  These routines do *not* use the float types from adt/.
+ *
+ *	  XXX These routines were not written by a numerical analyst.
+ *
+ *-------------------------------------------------------------------------
+ */
+#ifndef COMPLEX_H
```
Comment (hlinnaka) [inline src/include/utils/complex_type.h]: should be COMPLEX_TYPE.h for consistency

File: src/include/catalog/pg_operator.h [hunk 28]
```diff
@@ -981,6 +981,25 @@ DATA(insert OID = 3328 (  ">"     PGNSP PGUID b f f 3310 3310	16 3327 3329 gpxlo
...
+DATA(insert OID = 3472 (  "+"	   PGNSP PGUID b f f 193 193 193 0	0	complex_pl  - -));
+DATA(insert OID = 3473 (  "+"	   PGNSP PGUID l f f 0   193 193 0	0	 complex_up  - -));
+DATA(insert OID = 3474 (  "-"	   PGNSP PGUID b f f 193 193 193 0	0	 complex_mi  - -));
+DATA(insert OID = 3475 (  "-"	   PGNSP PGUID l f f 0   193 193 0	0	 complex_um  - -));
+DATA(insert OID = 3476 (  "*"	   PGNSP PGUID b f f 193 193 193 0	0	 complex_mul - -));
+DATA(insert OID = 3477 (  "/"	   PGNSP PGUID b f f 193 193 193 0	0	 complex_div - -)); 
+DATA(insert OID = 3478 (  "^"	   PGNSP PGUID b f f 193 193 193 0	0	 complex_power 	- -));
+DATA(insert OID = 3479 (  "|/"	   PGNSP PGUID l f f 0   193 193 0	0	 complex_sqrt 	- -));
+DATA(insert OID = 3480 (  "||/"	   PGNSP PGUID l f f 0   193 193 0	0	 complex_cbrt	- -));
+DATA(insert OID = 3481 (  "<"	   PGNSP PGUID b f f 193 193 16 3482 3484 complex_lt  scalarltsel scalarltjoinsel));
```
Comment (hlinnaka) [inline src/include/catalog/pg_operator.h]: Ok, why do we need < > operators for complex? There is no traditional concept of "greater than" or "less than" for complex numbers.

I think I know why: You need these operators to be able to create a b-tree index on complex types, don't you? Furthermore, you need B-tree opclass to do merge joins, for example. But for that, I think it would be more clear to name the operators something else, so that they're not confused with the traditional < >.

File: src/include/catalog/pg_operator.h [hunk 29]
```diff
@@ -981,6 +981,25 @@ DATA(insert OID = 3328 (  ">"     PGNSP PGUID b f f 3310 3310	16 3327 3329 gpxlo
 DATA(insert OID = 3329 (  "<="    PGNSP PGUID b f f 3310 3310	16 3330 3328 gpxloglocle scalarltsel scalarltjoinsel ));
 DATA(insert OID = 3330 (  ">="    PGNSP PGUID b f f 3310 3310	16 3329 3327 gpxloglocge scalargtsel scalargtjoinsel ));
 
+/* operators for complex data type */
+DATA(insert OID = 3469 (  "="	   PGNSP PGUID b t f 193 193 16 3469 3470 complex_eq eqsel eqjoinsel)); 
+#define ComplexEqualOperator 3469
```
Comment (hlinnaka) [inline src/include/catalog/pg_operator.h]: This define is unused

File: src/include/utils/builtins.h [hunk 30]
```diff
@@ -990,6 +990,7 @@ extern Datum numeric_demalg(PG_FUNCTION_ARGS); /* MPP */
 extern Datum int8_avg_demalg(PG_FUNCTION_ARGS); /* MPP */
 extern Datum float8_avg_demalg(PG_FUNCTION_ARGS); /* MPP */
 extern Datum numeric_avg_demalg(PG_FUNCTION_ARGS); /* MPP */
+extern Datum complex_cmp(PG_FUNCTION_ARGS);
```
Comment (hlinnaka) [inline src/include/utils/builtins.h]: Don't you need prototypes for the complex_lt, complex_gt, and other functions too?

File: src/include/catalog/pg_type.h [hunk 31]
```diff
@@ -410,7 +410,12 @@ DESCR("XML content");
 #define XMLOID 142
 DATA(insert OID = 143 ( _xml	   PGNSP PGUID -1 f b t \054 0 142 0 array_in array_out array_recv array_send - - - i x f 0 -1 0 _null_ _null_ ));
 
-/* OIDS 200 - 299 */
+DATA(insert OID = 193 (	complex	   PGNSP PGUID 16 f b t \054 0 0	194 complex_in complex_out complex_recv complex_send - - - d p f 0 -1 0 _null_ _null_ ));
+DESCR("double-precision floating point complex number, 16-byte storage");
+#define COMPLEXOID 193
+DATA(insert OID = 194 (	_complex	   PGNSP PGUID -1 f b t \054 0 193 0 array_in array_out array_recv array_send - - - d x f 0 -1 0 _null_ _null_ ));
+
+/*  OIDS 200 - 299  */
```
Comment (hlinnaka) [inline src/include/catalog/pg_type.h]: unnecessary whitespace change in the comment

File: src/include/catalog/pg_type.h [hunk 32]
```diff
@@ -410,7 +410,12 @@ DESCR("XML content");
 #define XMLOID 142
 DATA(insert OID = 143 ( _xml	   PGNSP PGUID -1 f b t \054 0 142 0 array_in array_out array_recv array_send - - - i x f 0 -1 0 _null_ _null_ ));
 
-/* OIDS 200 - 299 */
+DATA(insert OID = 193 (	complex	   PGNSP PGUID 16 f b t \054 0 0	194 complex_in complex_out complex_recv complex_send - - - d p f 0 -1 0 _null_ _null_ ));
```
Comment (hlinnaka) [inline src/include/catalog/pg_type.h]: OIDs 193 and 194 are used for different objects in later versions of PostgreSQL, so it'd be nice to renumber these to avoid a conflict in the future.

File: src/include/utils/float_utils.h [hunk 33]
```diff
@@ -0,0 +1,36 @@
+/*-------------------------------------------------------------------------
+ *
+ * complex.h - Declarations for complex data type
```
Comment (hlinnaka) [inline src/include/utils/float_utils.h]: this file is called float_utils.h

File: src/include/utils/float_utils.h [hunk 34]
```diff
@@ -0,0 +1,36 @@
...
+#define FLOAT_UTILS_H
+/*
+ * check to see if a float4/8 val has underflowed or overflowed
+ */
+#define CHECKFLOATVAL(val, inf_is_valid, zero_is_valid)			\
+do {															\
+	if (isinf(val) && !(inf_is_valid))							\
+		ereport(ERROR,											\
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),	\
+		  errmsg("value out of range: overflow"),errOmitLocation(true)));				\
```
Comment (hlinnaka) [inline src/include/utils/float_utils.h]: errOmitLocation not needed here either

File: src/backend/utils/adt/float.c [hunk 35]
```diff
@@ -43,23 +44,6 @@ static const uint32 nan[2] = {0xffffffff, 0x7fffffff};
 #define MAXFLOATWIDTH	64
 #define MAXDOUBLEWIDTH	128
 
-/*
- * check to see if a float4/8 val has underflowed or overflowed
- */
-#define CHECKFLOATVAL(val, inf_is_valid, zero_is_valid)			\
```
Comment (hlinnaka) [inline src/backend/utils/adt/float.c]: Since this is upstream code, it would be good to leave a comment here explaining that this has been moved to float_utils.h, and why. That'd help with future merges. (and a reverse comment in float_utils.h explaining where the code came from)

File: src/include/catalog/pg_amop.h [hunk 36]
```diff
@@ -914,7 +921,6 @@ DATA(insert (	3032	1700 1700 2 f 1755 3013 ));
 DATA(insert (	3032	1700 1700 3 f 1752 3013 ));
 DATA(insert (	3032	1700 1700 4 f 1757 3013 ));
 DATA(insert (	3032	1700 1700 5 f 1756 3013 ));
-
 /*
```
Comment (hlinnaka) [inline src/include/catalog/pg_amop.h]: spurious whitespace change

File: src/backend/utils/adt/complex_type.c [hunk 37]
```diff
@@ -0,0 +1,1385 @@
...
+}
+
+/*
+ *		complex_arg				- returns the phase of arg1
+ */
+Datum
+complex_arg(PG_FUNCTION_ARGS)
+{
+	Complex    *comp = PG_GETARG_COMPLEX_P(0);
+	double		result = 0;
```
Comment (hlinnaka) [inline src/backend/utils/adt/complex_type.c]: no need to initialize result to 0.

File: src/backend/utils/adt/complex_type.c [hunk 38]
```diff
@@ -0,0 +1,1385 @@
...
+	char	   *ascii = (char *) palloc(MAXCOMPLEXWIDTH + 1);
+	char	   *limit = ascii + MAXCOMPLEXWIDTH + 1;
+	char	   *p = ascii;
+
+	*p = '\0';
+	complex_part_out(&p, limit, re(num));
+	p += snprintf(p, limit - p, " %c ", (im(num) >= 0 || isnan(im(num))) ? '+' : '-');
+	
+	/*
+	 * Complex numbers with 0 imaginary part have different meaning wheter the imaginary part
```
Comment (hlinnaka) [inline src/backend/utils/adt/complex_type.c]: typo: wheter -> whether

**Comment by hlinnaka** [comment]
Left a bunch of nitpick line comments.

The < > operators need some thinking. Do we really should expose them to users with those names, and if not, what should they be called.

Having documentation ready be helpful for reviewing. For example, it'd make it easier to verify that the behaviour wrt. NaNs and infinities is consistent, and that the code implements it correctly, whatever we want the behaviour to be.

Aside from those things, LGTM.

File: src/backend/utils/adt/complex_type.c (hunk 21 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: I think this is based on the width of two float8 + padding. @macroyuyang can comment here please

File: src/include/catalog/pg_operator.h (hunk 28 above)
Comment (atris) [inline src/include/catalog/pg_operator.h]: Good point. Can we rename it to "complex_support_lt","complex_support_rt" etc?

**Comment by atris** [comment]
Merged,thanks!
